php-checks/src/main/java/org/sonar/php/checks/GenericExceptionCheck.java

```diff
@@ -48,87 +48,90 @@
 @BelongsToProfile(title = CheckList.SONAR_WAY_PROFILE, priority = Priority.MAJOR)
 @SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.EXCEPTION_HANDLING)
 @SqaleConstantRemediation("20min")
-public class GenericExceptionCheck extends SquidCheck<LexerlessGrammar> {
+public class GenericExceptionCheck extends PHPVisitorCheck {
 
+  public static final String KEY = "S112";
   public static final String MESSAGE = "Define and throw a dedicated exception instead of using a generic one.";
 
   private static final Set<String> RAW_EXCEPTIONS = ImmutableSet.of("ErrorException", "RuntimeException", "Exception");
-  private static final String SEPARATOR = PHPPunctuator.NS_SEPARATOR.getValue();
-  private String namespace;
-  private List<String> uses = Lists.newArrayList();
+  private Set<String> uses = Sets.newHashSet();
+  public String currentNamespace = "";
 
   @Override
-  public void init() {
-    subscribeTo(
-      PHPGrammar.NAMESPACE_STATEMENT,
-      PHPGrammar.USE_DECLARATION,
-      PHPGrammar.THROW_STATEMENT);
-  }
-
-  @Override
-  public void visitFile(@Nullable AstNode astNode) {
-    namespace = "";
+  public void visitScript(ScriptTree tree) {
     uses.clear();
+    currentNamespace = "";
+    super.visitScript(tree);
   }
 
   @Override
-  public void visitNode(AstNode astNode) {
-    if (astNode.is(PHPGrammar.NAMESPACE_STATEMENT)) {
-      parseNamespace(astNode);
-      return;
-    }
-    if (astNode.is(PHPGrammar.USE_DECLARATION)) {
-      parseUse(astNode);
-      return;
-    }
-    if (isBaseException(astNode)) {
-      getContext().createLineViolation(this, MESSAGE, astNode);
+  public void visitThrowStatement(ThrowStatementTree tree) {
+    NamespaceNameTree namespaceName = getNamespaceName(tree);
+
+    if (namespaceName != null && isGenericException(namespaceName)) {
+      context().newIssue(this.KEY, MESSAGE).tree(namespaceName);
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this instance-reference to a static reference. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2209)